File: app/src/main/java/com/example/mike/mp3player/service/library/content/retriever/SongsRetriever.java
package com.example.mike.mp3player.service.library.content.retriever;

import android.content.ContentResolver;
import android.database.Cursor;
import android.os.Handler;
import android.provider.MediaStore;
import android.support.v4.media.MediaBrowserCompat.MediaItem;

import androidx.annotation.NonNull;

import com.example.mike.mp3player.commons.MediaItemType;
import com.example.mike.mp3player.commons.MediaItemUtils;
import com.example.mike.mp3player.service.library.content.parser.ResultsParser;
import com.example.mike.mp3player.service.library.search.Song;
import com.example.mike.mp3player.service.library.search.SongDao;

import static com.example.mike.mp3player.service.library.content.Projections.SONG_PROJECTION;

public class SongsRetriever extends ContentResolverRetriever<Song> {

    public SongsRetriever(ContentResolver contentResolver,
                          ResultsParser resultsParser, SongDao songDao, Handler handler) {
        super(contentResolver, resultsParser, songDao, handler);
    }

    @Override
    public MediaItemType getType() {
        return MediaItemType.SONG;
    }

    @Override
    Cursor performGetChildrenQuery(String id) {
        return contentResolver.query(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, getProjection(),
                null, null, null);
    }


    @Override
    public String[] getProjection() {
        return SONG_PROJECTION.toArray(new String[0]);
    }

    @Override
    Song createFromMediaItem(@NonNull MediaItem item) {
        final String id = MediaItemUtils.getMediaId(item);
        final String value = MediaItemUtils.getTitle(item);
        return new Song(id, value);
    }

}

File: service/src/main/java/com/github/goldy1992/mp3player/service/library/content/retriever/SongsRetriever.kt
package com.github.goldy1992.mp3player.service.library.content.retriever

import android.content.ContentResolver
import android.database.Cursor
import android.provider.MediaStore
import com.github.goldy1992.mp3player.commons.MediaItemType
import com.github.goldy1992.mp3player.commons.dagger.scopes.ComponentScope
import com.github.goldy1992.mp3player.service.library.content.Projections.SONG_PROJECTION
import com.github.goldy1992.mp3player.service.library.content.parser.SongResultsParser

@ComponentScope
open class SongsRetriever constructor(contentResolver: ContentResolver,
                                      resultsParser: SongResultsParser)
    : ContentResolverRetriever(contentResolver, resultsParser, null) {
    override val type: MediaItemType
        get() = MediaItemType.SONG

    override fun performGetChildrenQuery(id: String?): Cursor? {
        return contentResolver.query(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, projection,
                null, null, null)
    }

    override val projection: Array<String?>?
        get() = SONG_PROJECTION.toTypedArray()
}